pkg/auth/providers/openldap/openldap_client.go

```diff
@@ -0,0 +1,447 @@
+	}
+	userPrincipal, groupPrincipals, err := p.getPrincipalsFromSearchResult(result, opResult, config, caPool)
+	if err != nil {
+		return v3.Principal{}, nil, nil, err
+	}
+
+	allowed, err := p.userMGR.CheckAccess(config.AccessMode, config.AllowedPrincipalIDs, userPrincipal, groupPrincipals)
+	if err != nil {
+		return v3.Principal{}, nil, nil, err
+	}
+	if !allowed {
+		return v3.Principal{}, nil, nil, httperror.NewAPIError(httperror.Unauthorized, "unauthorized")
+	}
+
+	return userPrincipal, groupPrincipals, map[string]string{}, err
+}
+
+func (p *openldapProvider) getPrincipalsFromSearchResult(result *ldapv2.SearchResult, opResult *ldapv2.SearchResult, config *v3.OpenLDAPConfig, caPool *x509.CertPool) (v3.Principal, []v3.Principal, error) {
+	var groupPrincipals []v3.Principal
+	var userPrincipal v3.Principal
+	var nonDupGroupPrincipals []v3.Principal
+
+	entry := result.Entries[0]
+	userAttributes := entry.Attributes
+	if !ldap.HasPermissionForOpenLDAP(userAttributes, config) {
+		return v3.Principal{}, nil, fmt.Errorf("Permission denied")
+	}
+
+	logrus.Debugf("getPrincipals: user attributes: %v ", userAttributes)
+
+	userMemberAttribute := entry.GetAttributeValues(config.UserMemberAttribute)
+	if userMemberAttribute == nil {
+		userMemberAttribute = opResult.Entries[0].GetAttributeValues(config.UserMemberAttribute)
+	}
+
+	logrus.Debugf("SearchResult memberOf attribute {%s}", userMemberAttribute)
+
+	isType := false
+	objectClass := entry.GetAttributeValues(ObjectClassAttribute)
+	for _, obj := range objectClass {
+		if strings.EqualFold(string(obj), config.UserObjectClass) {
+			isType = true
+		}
+	}
+	if !isType {
+		return v3.Principal{}, nil, nil
+	}
+
+	user, err := p.attributesToPrincipal(entry.Attributes, result.Entries[0].DN, UserScope, config)
+	userPrincipal = *user
+	if err != nil {
+		return userPrincipal, groupPrincipals, err
+	}
+
+	if userMemberAttribute != nil {
+		for _, dn := range userMemberAttribute {
+			query := "(&(" + config.GroupDNAttribute + "=" + dn + ")(" + ObjectClassAttribute + "=" +
+				config.GroupObjectClass + "))"
+			userMemberGroupPrincipals, err := p.searchLdap(query, GroupScope, config, caPool)
+			groupPrincipals = append(groupPrincipals, userMemberGroupPrincipals...)
+			if err != nil {
+				return userPrincipal, groupPrincipals, err
+			}
+		}
+	}
+
+	opEntry := opResult.Entries[0]
+	opAttributes := opEntry.Attributes
+
+	groupMemberUserAttribute := entry.GetAttributeValues(config.GroupMemberUserAttribute)
+	if len(groupMemberUserAttribute) == 0 {
+		for _, attr := range opAttributes {
+			if attr.Name == config.GroupMemberUserAttribute {
+				groupMemberUserAttribute = attr.Values
+			}
+		}
+	}
+
+	if len(groupMemberUserAttribute) > 0 {
+		query := "(&(" + config.GroupMemberMappingAttribute + "=" + groupMemberUserAttribute[0] + ")(" + ObjectClassAttribute + "=" +
+			config.GroupObjectClass + "))"
+		newGroupPrincipals, err := p.searchLdap(query, GroupScope, config, caPool)
```

Comment:
You pass in config and caPool here so that searchLdap can create a new connection, but you already have a connection. Can you refactor to use the connection you already have?